daemon/state.go

```diff
@@ -309,6 +309,9 @@ func (d *Daemon) regenerateRestoredEndpoints(state *endpointRestoreState) (resto
 
 	var endpointCleanupCompleted sync.WaitGroup
 	for _, ep := range state.toClean {
+		// Remove restored rules of cleaned endpoint
+		proxy.DefaultDNSProxy.RemoveRestoredRules(ep.ID)
```

Comment:
Should this be handled instead via either `Endpoint.Delete()` or `Endpoint.leaveLocked()`?

Reply:
Good point. That would cover the failed regeneration case as well, when the endpoint ends up being deleted eventually!

Reply:
Allegedly yes :-) I know we had a bug around this in the past where one path didn't properly clean up the endpoint but I *think* we converged the implementation towards actually calling the delete functions in each path. Maybe @christarazi recalls the details on what I'm alluding to here? I think it was related to metrics for endpoints on restore.

Reply:
Yes that's correct. The commit in question is here: https://github.com/cilium/cilium/pull/11884/commits/07a2171e64afebaecd0c2fc82cb8dae99b9d1723

Basically, if there was an attempt to create an endpoint with invalid data, the above commit now deletes that endpoint, rather than leaving it.

Reply:
WIll also have to move the removal after initial regeneration to happen from the endpoint after any successful regeneration to cover the case where the initial regeneration fails.

Reply:
Will this be done as a followup?